File: Grind75/updateMatrix.js
function updateMatrix(mat) {
  const rows = mat.length;
  const cols = mat[0].length;
  const queue = [];

  // Initialize the matrix: 0 stays 0, 1 becomes Infinity
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (mat[r][c] === 0) {
        queue.push([r, c]);
      } else {
        mat[r][c] = Infinity;
      }
    }
  }

  const directions = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1],
  ];

  // BFS from all 0s
  while (queue.length > 0) {
    const [r, c] = queue.shift();

    for (const [dr, dc] of directions) {
      const newRow = r + dr;
      const newCol = c + dc;

      if (
        newRow >= 0 &&
        newRow < rows &&
        newCol >= 0 &&
        newCol < cols &&
        mat[newRow][newCol] > mat[r][c] + 1
      ) {
        mat[newRow][newCol] = mat[r][c] + 1;
        queue.push([newRow, newCol]);
      }
    }
  }

  return mat;
}

File: Grind75/update_matrix.py
def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
    rows, cols = len(mat), len(mat[0])
    queue = deque()
    dist = [[float('inf')] * cols for _ in range(rows)]

    for row in range(rows):
        for col in range(cols):
            if mat[row][col] == 0:
                dist[row][col] = 0
                queue.append((row, col))

    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

    while queue:
        row, col = queue.popleft()

        for dr, dc in directions:
            newRow, newCol = row + dr, col + dc

            if 0 <= newRow < rows and 0 <= newCol < cols:
                if dist[newRow][newCol] > dist[row][col] + 1:
                    dist[newRow][newCol] = dist[row][col] + 1
                    queue.append((newRow, newCol))

    return dist